### PR TITLE
Added tiledmap loader for JSON files. Replacement for PR (#7386)

### LINKED ISSUE #5707: [Feature Request] Load TiledMap from Json
**Issue details**

With Libgdx, we can create a TiledMap from a TMX file.
Is it conceivable to create a TiledMap from a JSON file ?
This could be useful for people who want to serialize all their game in JSON.

**Version of LibGDX and/or relevant dependencies**
GDX 1.9.9

### LINKED ISSUE #7291: TiledMap can be loaded only in xml format and not json.
Basically this is a request for new features. 

I would like to have the chance to load a map in json (.tmj) format from TILED. Right now only .tmx formats are loaded. 
More i would like to have in the loader also a method to load a defined Tileset file (both formats as well) directly into a TiledMapTileSet. 

Thanks a lot for your efforts.
Keep up the incredible good work.

### PR DESCRIPTION
**Pull request by BoBIsHere86**
**The Purpose of the PR:**
Add a loader for TiledMaps which allows users to load maps in the JSON (.tmj) format.

This PR also closes Issues: fixes #5707 and fixes #7291

**This PR is meant to replace #7386**
All credit for the original TmjLoader code should go to @Kevin-Garnier.
The original PR wasn't quite finished. There were a lot of changes that needed to be made to the original code and after reaching out to him, he said he no longer had the time to work on this. So he won't able to make any of my requested changes and said we can continue on without him.

I have fixed all of the oversights and bugs I came across when testing his original code. These can be found in the comments in the original PR (#7386).

In addition to fixing the bugs I came across in the Original PR. I added a new test for loading JSON(.tmj) files.
6 of the 7 test maps were converted from the old staple .tmx maps we use in most of our tests and I added a 7th one which demonstrates some extra features, image layers, animated tiles, etc. Reusing the same tilesets we have access to.
Test seen below:
jsonTest

Since there is a shortage of people who can look over TiledMap PRs.
I will be pinging @Quillraven, (No Pressure!)
If he has any spare time, maybe he can give this code a quick once over at some point, since he has helped me on other tiledmap PRs.

### CODE REVIEW

**Comment by Quillraven** [comment]
Did not have time yet to test the changes but from a quick look on my phone, if I see that correctly, there is a separate loader for tmj and tmx and the user has to choose which one he wants?

I find that inconvenient from a user point of view. Imo a user wants a loader for "Tiled" editor maps and it should just work. No matter if he stores in tmx or tmj format. The loader itself should know what to do most likely depending on file ending or header?

Also, in case someone uses tmx and tmj in the same project, he has to switch loaders? Does it even work to load a tmx and tmj map at the same time if they are both queued in the Asset Manager?

I try to test the changes and have a look by end of next week latest.

**Comment by BoBIsHere86** [comment]
> Did not have time yet to test the changes but from a quick look on my phone, if I see that correctly, there is a separate loader for tmj and tmx and the user has to choose which one he wants?
> 
> I find that inconvenient from a user point of view. Imo a user wants a loader for "Tiled" editor maps and it should just work. No matter if he stores in tmx or tmj format. The loader itself should know what to do most likely depending on file ending or header?
> 
> Also, in case someone uses tmx and tmj in the same project, he has to switch loaders? Does it even work to load a tmx and tmj map at the same time if they are both queued in the Asset Manager?
> 
> I try to test the changes and have a look by end of next week latest.

Awesome, this is exactly why I pinged you. 👍 

I'll try to answer your questions to the best of my ability. (keep in mind I was not the one who made this, I just fixed all the errors to get it in a working state and added a test)

So yes, it seems the way it was designed was to take the path of least resistance and copy the way we load in the Tmx files.
For example instead of TmxMapLoader, a user would do something like this.
```
assetManager.setLoader(TiledMap.class, new TmjMapLoader(new InternalFileHandleResolver()));
```

I can kind of understand his reasoning for doing this. Since the only map loader we had, was specifically Titled "TmxMapLoader".
It makes sense he made a new one for specifically for Tmj files. Now a universal MapLoader would of course be optimal. But would require a bit of refactoring and testing.

Also, I didn't even think of a situation where a user might want to use both map types. My guess is that would not work. Since each would be tied to the TiledMap.class in the loader. So whichever was set last would be the one used.

I'll take a look into this later as well.

Thanks again, no rush.

**Comment by tommyettinger** [comment]
I have very little experience with AssetManager code, but I have made loaders that can be given either of two filetypes when loading fonts. The thing is, my code worked with Skin mainly, and I don't know how different it would be with primarily AssetManager. There is AssetManager code in the same folder as this, though: https://github.com/tommyettinger/textratypist/blob/main/src/test/java/com/github/tommyettinger/fontwriter/JsonSkin.java#L78 That folder is in tests because I haven't pulled it out into its own library yet, but it loads BitmapFont instances from either the usual .fnt files or the new .json/.dat files (where .dat here is just compressed .json using a GWT-compatible method). That's similar to loading maps from .tmx or .tmj , right?

**Comment by BoBIsHere86** [comment]
> I have very little experience with AssetManager code, but I have made loaders that can be given either of two filetypes when loading fonts. The thing is, my code worked with Skin mainly, and I don't know how different it would be with primarily AssetManager. There is AssetManager code in the same folder as this, though: https://github.com/tommyettinger/textratypist/blob/main/src/test/java/com/github/tommyettinger/fontwriter/JsonSkin.java#L78 That folder is in tests because I haven't pulled it out into its own library yet, but it loads BitmapFont instances from either the usual .fnt files or the new .json/.dat files (where .dat here is just compressed .json using a GWT-compatible method). That's similar to loading maps from .tmx or .tmj , right?

Yea, I'm not too well versed in the overall workings of the AssetManager.
But when it comes to the tiled mapping files, the .tmj and .tmx files need to be parsed and handled in an entirely separate way.
There's even some inconsistencies between properties contained in the exact same map but saved in different formats. (I learned this the hard way when debugging the original PR)
Like In the .tmj format, it explicitly denotes String properties as type="string" while in the .tmx format, there's no type at all. MapObjects in the .tmj file don't include a type specifying them as such, but in the .tmx they do.... little things like that.

The simplest solution to @Quillraven's suggestion about handling both file types would probably be to just make a new Generic "TiledMapLoader" class which could just act as a simple wrapper that just checks the file extension and then passes it on to the  appropriate loader. This way we could leave the TmxMapLoader untouched and people who've been using it for the past decade as well as all the libGDX Tests, can continue to use it without us causing a breaking change.

This of course, leads to a bigger discussion on the tiledmap code in general. In major need of a refactoring.
But that's why I've been trying to clean up these PRs recently. Most are pretty close to completion and once we can get everything fixed and up to date. We can take a step back and go. Ok, now let's clean this mess up... lol

**Comment by patton73** [comment]
My opinion on this is that for sure using mixed xml/json won't do the job. Freedom is cool but this is basically mess (i mean allowing users/programmers doing that is basically useless and counterproductive). I would stay with separated loaders. If you want to add something i would propose to add a factory that creates the parser given a parameter.
Keep up the good work.

Thanks.

**Comment by BoBIsHere86** [comment]
> My opinion on this is that for sure using mixed xml/json won't do the job. Freedom is cool but this is basically mess (i mean allowing users/programmers doing that is basically useless and counterproductive). I would stay with separated loaders. If you want to add something i would propose to add a factory that creates the parser given a parameter. Keep up the good work.
> 
> Thanks.

I can definitely see this side of things as well. I personally think someone will either be using one format or the other. And not mixing and matching.
Of course I can definitely see where Quillraven is coming from. I'm not against the simplicity of just having a single loader which handles everything. But I also don't want to side track this PR. We can always come back and fix the loader code and make it more robust at a later point.

***Also**, I would like to point out, and this slipped my mind since I've never used it.
We already have another loader called "AtlasTmxMapLoader". Which is a loader for tmx map files that use a texture atlas.
So technically we already leave it up to the use to correct loader already. So adding the TmjMapLoader, might not be that big of a deal.  Just my thoughts.

**Comment by tommyettinger** [comment]
Does Tiled allow you to open a .tmj map and save it as a .tmx map? If it does, I'm not even sure what the advantages might be to using .tmj (i.e. why we should partially support it, but not with atlases). The font loader I linked is needed for my purposes because a tool I use heavily generates a custom JSON format for fonts and can't generate .fnt (AngelCode BMFont) files. If Tiled can exchange .tmj and .tmx files... there isn't much reason to add support for .tmj, especially if it adds complexity to the several other PRs in process.

**Comment by Quillraven** [comment]
I just checked it because I also never used it. Always used xml so far and didn't know that there are multiple formats:

image

There is even a third one :D

My though process was more like:
- What if someone has only used XML until now because JSON wasn't supported and now he wants to try out JSON. In this scenario someone needs to convert all his XML maps to JSON first
- What if JSON or XML format has a bug in Tiled (e.g. a property is wrongly exported or whatever). In that case you must use whatever format is working in worst case you end up with XML and JSON maps and that wouldn't work so nicely with our approach here

----

But I agree. It is most likely not a (big) issue for the LibGDX Tiled users so we can stick to the existing approach and just support one Tiled loader for a specific format at a time.

**Comment by BoBIsHere86** [comment]
I can't really find any solid reasonings to choose one format over the other. But we have 2 active issues with users specifically requesting it. So there does seem to be people who want use that format. 

Also, I'd imagine the best reason to give support the .tmj format. Would be for users migrating from other engine's or frameworks.
Let's say someone decided to ditch Unity and come on down to libGDX town. But they have 50+ tiled maps all in .tmj format. Yea, I guess they could manually go and convert each one, one by one. Or we could just support it as well.

**Comment by tommyettinger** [comment]
They could also have them in Tiled's .js format, which there is no effort to support. This seems like a better use case for a bulk converter that takes .tmj (plus maybe .js) files and makes .tmx files, than to add (partial) support for .tmj ... The lack of atlas support for .tmj , to me, means it isn't as suitable for games that need better performance. There's also going to be more maintenance effort to support two formats (and not support a third, the .js one) as Tiled inevitably makes changes on their side.

**Comment by patton73** [comment]
> I can't really find any solid reasonings to choose one format over the other. But we have 2 active issues with users specifically requesting it. So there does seem to be people who want use that format.
> 

Well one reason is that xml is one of the worst and over engineered exchange data format ever created. And is what still remain of 90's and 2000's crap engineering. Very likely json parsing is faster since the format is the simplest you can imagine.

And i won't be suprised if some day Tiled will ditch xml in favor of json (that actually is the today's standard in data inter change). 

Honestly i cannot understand why the have also the third standard. That seems to me a javascript format. 

But that's it.

**Comment by Quillraven** [comment]
> They could also have them in Tiled's .js format, which there is no effort to support. This seems like a better use case for a bulk converter that takes .tmj (plus maybe .js) files and makes .tmx files, than to add (partial) support for .tmj ... The lack of atlas support for .tmj , to me, means it isn't as suitable for games that need better performance. There's also going to be more maintenance effort to support two formats (and not support a third, the .js one) as Tiled inevitably makes changes on their side.

To be fair does anyone know how the atlas loader works? I tried it once and invested some time but I never got it to work.

My workflow is to create the atlas myself and use it for the tileset/map. This is quite easy with tile extruder or texture packer gui.

----

However, I like your idea of "just" writing a converter. I'd prefer JSON though as the leading format over XML but that is just my personal preference.

We then just have one loader/parse logic (e.g. JSON) and the other loaders just convert to it. Maybe there is even a utility method or library from Tiled itself to easily convert between formats? Then we don't need to write our own (although I think it should not be that difficult because both XML and JSON are very popular and common. Pretty sure there are converters already available).

**Comment by tommyettinger** [comment]
Ideally, Tiled or its community already has such a tool written. Otherwise, if there's a way to deserialize a .tmj and load it into memory, then serialize into a .tmx , that pretty much is all that's needed. This PR may contain most or all of the loading code, I just don't know how complex saving is or whether it's been implemented already. I think to make compatible .tmx files, we wouldn't be able to just use the XmlWriter class in utils.

**Comment by BoBIsHere86** [comment]
Great discussion we have going on here. But there are some points I would like to bring up after some more digging.

**First, on the TiledMap JSON format in general** according to the documentation.
"The JSON format is the most common additional file format supported by Tiled. It can be used instead of TMX since Tiled can also open JSON maps and tilesets and the format supports all Tiled features. Especially in the browser and when using JavaScript in general, the JSON format is easier to load."

**Second the AtlasTmxMapLoader.** This isn't a Tiled specific map type and in order to use it you need to add a map-level property called 'atlas' with its value being the relative path to the TextureAtlas. I've never used it but I (Think) it's meant to be used with the [libgdx-tiledmappacker](https://github.com/tommyettinger/libgdx-tiledmappacker) As a quick way to combine your tilesets assets if u are using multiple ones across all your maps?
But I think this only supports .tmx files as it is. But it's a custom libGDX thing someone added along the way.

Normally when I create a tiledmap, I have all my tiles in a single atlas/spritesheet anyway. So there was never a need to look into using this. So my all map related rendering only ever took a single draw call. 

**Third the .js format**.
So the .js format was kind of a mystery to me until I started searching for it in the official Repo.
It's not specifically mentioned in the TiledMap documentation like .tmj or .tmx. But it's basically just the json wrapped in a function call.
It seems like It was added specifically at the request of javascript devs based on this issue from 2013 
 https://github.com/mapeditor/tiled/issues/374 and later merged from this PR https://github.com/mapeditor/tiled/pull/402
 
I still think at the bare minimum. We should attempt to support the main 2 formats that Tiled's been using for the past decade or so.

**Comment by Quillraven** [comment]
fyi I also in Tiled Discord about the conversion. Here is the answer:

> You should not support the "JS" format, it's meant for specific needs and not as a general-purpose Tiled format.
>
> There's no minimalistic Tiled utility for such a conversion. Tiled itself can do it (including via CLI), but this entails loading the entire map and then saving it as the other format.
>
> Unless you intend to write some sort of text-transforming conversion, anything you do will likely entail JSON parsing, and in that case I feel it's best to write a proper parser, even if it means having redundant code or refactoring the existing code to pull the parts that generate the user-facing representations into reusable functions.

**to summarize**:
- JS format is not needed. We should ignore that
- Tiled could convert it including CLI but imo that is not useful for us

----

So what do we do with this PR now? Do we agree to keep the existing XML loaders (normal and Atlas) and use them also for tmj? For tmj we write a converter tmj -> tmx?

**Comment by patton73** [comment]
I personally do not like the idea of an external converter (and it is not needed since Tiled can do it itsself, just load json map and save it in xml).  I would like to have a json loader. Or something similar. I mean converting the maps to json from xml is an useless waste of time since i can use the xml loader without converting the map.

The loader instead adds a feature so i can save maps in Tiled directly in json format and load it with libgdx.

My two cents.

**Comment by BoBIsHere86** [comment]
So I didn't have to much time to dive into this. But I was curious about any possible benefits to the json format.
So don't take this as a definitive test.  But i modified some game jam code which had a tiled map loader and hacked together a quick method to calculate average loading times. I turned one of my maps into a large monstrosity, 320 by 180 map size, 50+ tile and object layers, and tiles. I created a tmj and tmx version with each supported compression. Also to note, the TiledMap program itself begins to stutter when having this file open.

**First:** You can see the first difference is the .tmj file Sizes always end up slightly larger than the .tmx ones.
sizecomparison

**Second:**  This was the very quickly slapped together code to get a baseline to test the loading times:
```
public long loadBigTmj(){

		// Start timing the entire loading process
		long overallStartTime = System.nanoTime();

		if(assetManager.isLoaded(AssetDescriptors.tmjBIG)) {
			assetManager.unload(AssetDescriptors.tmjBIG.fileName);
		}

		assetManager.setLoader(TiledMap.class, new TmjMapLoader(new InternalFileHandleResolver()));
		assetManager.load(AssetDescriptors.tmjBIG);
		assetManager.finishLoading();


		// End timing for the overall loading process
		long overallEndTime = System.nanoTime();
		long totalTimeMillis = (overallEndTime - overallStartTime) / 1_000_000;
		System.out.println("Total loading time for BIG TMJ was " + totalTimeMillis + " ms");

		return totalTimeMillis;
	}
```
I made an equivalent for the TMX files as well.

I then called these methods 100 times each in a loop after a 5 iteration warm up. here
```
  public void testTMJLoadingTimes() {
        int iterations = 100;
        long totalLoadingTime = 0;

        // Optional: Warm-up runs
        int warmUpRuns = 5;
        for (int i = 0; i < warmUpRuns; i++) {
            assets.loadBigTmx();
        }

        // Actual timing runs
        for (int i = 0; i < iterations; i++) {
            System.out.println("Iteration " + (i + 1) + " of " + iterations);

            long loadingTime =  assets.loadBigTmj();
            totalLoadingTime += loadingTime;
        
        }

        double averageLoadingTime = (double) totalLoadingTime / iterations;
        System.out.println("Average loading time over " + iterations + " runs: " + averageLoadingTime + " ms");
    }

```
I would run one map type. stop, get results, then change code to do another map type. start again and repeat.
The results were below:

**100 runs, after 5 run warmup :**
**CSV parsing**
tmx: 211.76 ms
tmj: 424.01 ms

**Base64 uncompressed**
tmx: 282.65 ms
tmj: 177.72 ms

**zlib**
tmx: 113.76 ms
tmj: 108.36 ms

**gzip**
tmx: 103.27 ms
tmj: 98.07 ms

So for what it's worth the results were pretty interesting. With the Exception of when parsing CSVs.
The .tmj maps  were always slightly faster, most noticeably when handling base64 uncompressed. (which is good considering people are forced to use non compressed maps if you plan to do an html build)

Not sure why tmj has an issue with the csv parsing. my guess is it has something to do with the Json parser. 
Anyway I know this was not the most comprehensive or scientific benchmark. But at least it's something. lol

**Comment by BoBIsHere86** [comment]
We've had some great momentum on this PR and I appreciate everyone pitching in.

**So where do we go from here?**
(Warning! Some Personal Opinions Below.)
Quick Summary, taking into consideration everyone's input:

The TiledMap editor as it is now supports 2 file formats for exporting maps:
   1. A XML format which includes .tmx (map files), .tsx(tileset files) and .tx(template object Files)*
   2. A JSON format which includes .tmj (map files), .tsj(tileset files) and .tj(template object Files)*
   *Template objects are currently not supported in either format, but I have a feature PR ready to go when it's time.

These 2 formats have been used by Tiled for the last 10 years or so. Checking through upcoming project milestones on github. There doesn't seem to be any inclination this will change anytime soon.

We current only support the XML format through the TmxMapLoader and AtlasTmxMapLoader*.

First, should we support the JSON format?
**Positives:**
1. Personal testing shows, on average the .tmj files when saved under Base64 encoding load faster than the .tmx ones.
2. Multiple requests from users have been made asking to support of this format.

**Negatives:**
1. The size of equivalent .tmj files when compared to .tmx are a bit larger.
2. No equivalent "AtlasTmjMapLoader" for json files***
**Small Tangent About AtlasTmxMapLoader:** 
*I finally took some time to dive into the AtlasTmxMapLoader. And I have a pretty good grasp on it's purpose now. 
-A quick rundown for those who havn't used it. It seems to be a "nice to have" feature through libGDX. If a user happens to be using multiple tilesheets/tilesets for their maps, this tool can be run to combine them all into 1 big tilesheet atlas instead of being split across multiple tilesheets. BUT this is a one way transaction. The resulting map and atlas wont be openable without errors in the TiledMap editor once modified. So this is meant to be done at the end of your workflow once you save your map. If you need to change your map you would have to redo this step before bringing it back into libGDX.
Personally, I have never needed to use this, as I have never had a situation where I've had multiple tilesheets being used across a single map. BUT I do recognize that it's a handy feature for those who could benefit from it.

So should we have an AtlasLoader for the JSON format as well, a AtlasTmjMapLoader if you will?
Yes! Absolutely, why not? The more options the better. The only problem though is that the libgdx-tiledmappacker only supports .tmx file right now. BUT, that's because the libgdx-tiledmappacker relies on the TmxMapLoader to read and build out the new .tmx map file. So if we decide to go ahead and support the .tsj format. Then we can work on adding this as a future enhancement.
***Bonus Tangent***
I did come across some issues with the AtlasTmxMapLoader. If you have a map with any ImageLayers you can't use the AtlasTmxMapLoader, it will crash. (I have no idea how no one has reported this yet) This is because it has it's own ImageResolver implementation, DirectAtlasResolver and AssetManagerAtlasResolver. When getImage() is called it attempts to looking inside the Atlas for the images. BUT libgdx-tiledmappacker does NOT add imagelayer images to it's atlas(this would actually be super useful and would make me use it). So it just breaks in the BaseMapLoader calls.

Now, I understand it was suggested to maybe have another outside tool to convert from .tmj to .tmx. But then you lose out on the benefit of the faster parsing and adds an un-needed extra step.
Also, the idea of a universal loader. Which does sound great but that is also something which could be added as an enhancement later on. (or in the much needed overhaul one day)

So where does that leave us? 
Request changes? Move forward? Banish this PR to hell? Let us know.

*Thanks for all the help and input so far!

**Comment by patton73** [comment]
According to me:

1) if the merge of this PR is a problem because then the code must be mantained (and this become a problem) then do not merge it. XML is still there and at least we can use it.
2) **please do not forget and merge the part that makes public the tileset loader methods in xml reader.** This is vital to make random maps and use tiled's generated tilesets.
3) Please merge all other PRs regarding tilemap's bugs.

Thanks for all the effort and good intentions.

**Review by Quillraven — CHANGES_REQUESTED** [review]

File: gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java
Comment (BoBIsHere86) [reply gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java:167]: Oh yea, that's the big issue I had with this class as well. It IS literally a copy paste. But using the JSON parser instead of the XML one and there a small differences in the properties between the json format and the xml.
So should we keep it small and build out a utility class for some shared functionality?
Or do a larger refactor. I feel as though having BaseTmxMapLoader and BaseTmjMapLoader having a shared parent class would make sense? A BaseMapLoader class which they both could extend? But that seems like it's leaking outside of the original intent of the PR.
What do you think? Keep it small?

File: gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java
Comment (BoBIsHere86) [reply gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java:571]: Noting this as a change for both map loaders. Possible inclusion into a utility class or larger refactor into a parent class which both loaders extend.

File: gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java
Comment (BoBIsHere86) [reply gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java:36]: Noting this as a change for both map loaders. Possible inclusion into a utility class or larger refactor into a parent class which both loaders extend.

File: gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java
Comment (BoBIsHere86) [reply gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java:551]: Ok, quick question.
So the tmx and tmj formats differ in this regard.
When tiled generates the TMJ file it explicitly denotes String properties as type="string" while in the TMX file there is no type added and that denotes a it as string, so in the castProperty() method for the TMX loader they look for nulls 
to mean String property.

So 2 options. 
1. We check for null in the TMJ loader and give the user the same "unsupported Type Message"
OR
2. We can cover both cases by doing this `if (type == null || "string".equals(type))  ` and then this can be another method added to our new refactored utility or parent class.

File: gdx/src/com/badlogic/gdx/maps/tiled/TmjMapLoader.java
Comment (BoBIsHere86) [reply gdx/src/com/badlogic/gdx/maps/tiled/TmjMapLoader.java:256]: Maybe this is another candidate for a refactor, they all use the same parameters.
Wow... I am having the same issue. Locally I see
```
 public static class Parameters extends AssetLoaderParameters<TiledMap> {
		  /** generate mipmaps? **/
		  public boolean generateMipMaps = false;
		  /** The TextureFilter to use for minification **/
		  public TextureFilter textureMinFilter = TextureFilter.Nearest;
		  /** The TextureFilter to use for magnification **/
		  public TextureFilter textureMagFilter = TextureFilter.Nearest;
		  /** Whether to convert the objects' pixel position and size to the equivalent in tile space. **/
		  public boolean convertObjectToTileSpace = false;
		  /** Whether to flip all Y coordinates so that Y positive is up. All libGDX renderers require flipped Y coordinates, and thus
			* flipY set to true. This parameter is included for non-rendering related purposes of tmj files, or custom renderers. */
		  public boolean flipY = true;
	 }
```
But github is empty....

File: gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java
Comment (Quillraven) [reply gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java:167]: I think keep it as it is for now. But should be refactored in the future at some point.

File: gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java
Comment (Quillraven) [reply gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java:551]: I vote for option 2

File: gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java
Comment (BoBIsHere86) [reply gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java:167]: Alright, Sounds like a plan! I actually started a refactor locally. But then realized if we made such sweeping changes right now it's going to have big consequences for all of the 8+ tiled PRs that are waiting to get merged. Once everything is stable we can worry about a large refactor.

Thanks for the input!

Comment (BoBIsHere86) [reply gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java:167]: Ok, so one last follow up. Need your opinion. 
So I'm building out a small utility class which should clean up a bunch of duplicated methods that are not specific to XML vs Json parsing.
```
public class TiledMapLoaderUtils {

	 protected static final int FLAG_FLIP_HORIZONTALLY = 0x80000000;
	 protected static final int FLAG_FLIP_VERTICALLY = 0x40000000;
	 protected static final int FLAG_FLIP_DIAGONALLY = 0x20000000;
	 protected static final int MASK_CLEAR = 0xE0000000;

	 public static Object castProperty (String name, String value, String type) {
		  if (type == null || "string".equals(type)) {
				return value;
		  } else if (type.equals("int")) {
				return Integer.valueOf(value);
		  } else if (type.equals("float")) {
				return Float.valueOf(value);
		  } else if (type.equals("bool")) {
				return Boolean.valueOf(value);
		  } else if (type.equals("color")) {
				// Tiled uses the format #AARRGGBB
				String opaqueColor = value.substring(3);
				String alpha = value.substring(1, 3);
				return Color.valueOf(opaqueColor + alpha);
		  } else {
				throw new GdxRuntimeException(
					"Wrong type given for property " + name + ", given : " + type + ", supported : string, bool, int, float, color");
		  }
	 }

	 public static Cell createTileLayerCell (boolean flipHorizontally, boolean flipVertically, boolean flipDiagonally) {
		  Cell cell = new Cell();
		  if (flipDiagonally) {
				if (flipHorizontally && flipVertically) {
					 cell.setFlipHorizontally(true);
					 cell.setRotation(Cell.ROTATE_270);
				} else if (flipHorizontally) {
					 cell.setRotation(Cell.ROTATE_270);
				} else if (flipVertically) {
					 cell.setRotation(Cell.ROTATE_90);
				} else {
					 cell.setFlipVertically(true);
					 cell.setRotation(Cell.ROTATE_270);
				}
		  } else {
				cell.setFlipHorizontally(flipHorizontally);
				cell.setFlipVertically(flipVertically);
		  }
		  return cell;
	 }

	 public static FileHandle getRelativeFileHandle(FileHandle file, String path) {
		  StringTokenizer tokenizer = new StringTokenizer(path, "\\/");
		  FileHandle result = file.parent();
		  while (tokenizer.hasMoreElements()) {
				String token = tokenizer.nextToken();
				if (token.equals(".."))
					 result = result.parent();
				else {
					 result = result.child(token);
				}
		  }
		  return result;
	 }

	 public static void addStaticTiledMapTile(TiledMapTileSet tileSet, TextureRegion textureRegion, int tileId, float offsetX, float offsetY, boolean flipY) {
		  TiledMapTile tile = new StaticTiledMapTile(textureRegion);
		  tile.setId(tileId);
		  tile.setOffsetX(offsetX);
		  tile.setOffsetY(flipY ? -offsetY : offsetY);
		  tileSet.putTile(tileId, tile);
	 }

	 public static int unsignedByteToInt (byte b) {
		  return b & 0xFF;
	 }

protected void addStaticTiledMapTile (TiledMapTileSet tileSet, TextureRegion textureRegion, int tileId, float offsetX,
		float offsetY) {
		 TiledMapLoaderUtils.addStaticTiledMapTile(tileSet, textureRegion, tileId, offsetX, offsetY, flipY);
	}
```

But now here's a question, almost all of these methods in the BaseTmxMapLoader were set as protected.
Examples: "protected Cell createTileLayerCell()","protected Object castProperty()","protected static FileHandle getRelativeFileHandle()","protected static int unsignedByteToInt()"

So proper convention would say we shouldn't get rid of these in the BaseTmxMapLoader right? Just in case there might be someone out there using this codebase with their own implementation, like a custom map loader or something?
Which would mean we could do this.
```
protected static int unsignedByteToInt (byte b) {
		return TiledMapLoaderUtils.unsignedByteToInt(b);
}

protected static FileHandle getRelativeFileHandle (FileHandle file, String path) {
		 return TiledMapLoaderUtils.getRelativeFileHandle(file, path);
}
```
Personally, this looks like ugly clutter for such small changes and in the libGDX codebase they are only being called by the child classes of the Tmx loader which I would update to use the new utility methods anyway. But I don't know, should we even be worried this? Am I overthinking these small changes?

Comment (Quillraven) [reply gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java:167]: I am personally not a big fan of `utils` classes but I am also not a big fan of a lot of inheritance levels either :D But since we won't make it perfect in this PR I'd go for a base class as you initially mentioned. That way you keep the methods and people who are overriding it don't get a broken code.

Comment (BoBIsHere86) [reply gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java:167]: Ok cool. That probably is the way to go.
So my plan of attack would be to start with a universal parent class which both the BaseTmxMapLoader and BaseTmjMapLoader can extend and derive all their currently duplicated methods and fields from. This way we can keep all relevant map related methods together and gives us a place for a much larger refactor in the future.

**Comment by BoBIsHere86** [comment]
I have completed the small refactor which Quillraven and I discussed. This eliminated a ton of duplicate code throughout the MapLoader classes, the base map loader classes now extend the parent BaseTiledMapLoader. This provides the implementation of all shared methods currently used by all the maploaders without  breaking code for any user who's previously extended the TmxMapLoader classes.
BaseTiledMapLoader now contains:
getIdToObject(), castProperty(), createTileLayerCell(),unsignedByteToInt(),getRelativeFileHandle(),addStaticTiledMapTile() as well as the Parameters, and field variables used by both subclasses. Giving us a place for all future tilemap loading specific methods to be shared.

I have also gone ahead and added one more json specific test related to object loading. TiledMapJsonObjectLoadingTest.
This is a replica of the TiledMapObjectLoadingTest.
Below you can see them side by side, looking just as hideous as ever.
jsontest2


***Removal of the loadCustomTileSet() Methods:**
Originally there was a second request, piggybacking along with this one, that was held over form the original PR, unrelated to JSON support.
I had a quick chat with the requester and the results of the proposed loadCustomTileSet() that had been implemented was not what he had wanted. As to not delay this PR any longer and keep it specific to it's original intent, (Adding json support). These additions have been removed and that request should be placed in it's own issue at a later date.

**Review by Quillraven — APPROVED** [review]

**Comment by BoBIsHere86** [comment]
James brought to my attention a tileset file that should be renamed, sooner rather than later. I thought it best to sneak it in here,  as this PR should be merged next.

...and now that that's out of the way I think we can finally close the book on this whole JSON support fiasco and move forward. This was really the only big PR holding us up. The next ones should all be a piece of cake.

Thanks for all the help!  :tada:

**Review by tommyettinger — APPROVED** [review]
This PR looks huge and incredibly thorough. I like that it only makes changes to the maps package; that drastically reduces the odds of a bug drifting out into the code at-large. It's practically my mantra at this point, but I don't use Tiled, so there's not much I can say at this point. Even though it seems fairly easy to convert .tmj to .tmx using Tiled itself, I guess people keep requesting .tmj support... This seems to work, so I think merging seems like a good idea.

**Comment by Quillraven** [comment]
Who can merge PRs?

**Comment by tommyettinger** [comment]
I can, though I should really wait for a second approval. Not sure who else even reads Tiled PRs...

**Review by SimonIT — APPROVED** [review]
LGTM, fixes #5707 and fixes #7291

**Pull request merged**
